libqpdf/QPDF.cc

```diff
@@ -1292,48 +1292,48 @@ QPDF::showXRefTable()
     }
 }
 
-// Ensure all objects in the pdf file, including those in indirect references,
-// appear in the object cache.
 void
 QPDF::fixDanglingReferences(bool force)
 {
+    // Ensure all objects in the pdf file, including those in indirect
+    // references, appear in the object cache.
     if (this->m->fixed_dangling_refs && !force) {
         return;
     }
 
-    if (!this->m->fixed_dangling_refs) {
-        // First pass is only run if the the xref table has not been
-        // reconstructed. It will be terminated as soon as reconstruction is
-        // triggered.
-        if (!this->m->reconstructed_xref) {
-            for (auto const& iter: this->m->xref_table) {
-                auto og = iter.first;
-                if (!isCached(og)) {
-                    m->obj_cache[og] =
-                        ObjCache(QPDF_Unresolved::create(this, og), -1, -1);
-                    if (this->m->reconstructed_xref) {
-                        break;
-                    }
-                }
-            }
```

Comment:
How about having a resolve loop after the for loop if reconstructed_xref == false.

It after the the resolve loop reconstructed_xref is still false we can return, otherwise we continue with the second pass.

Reply:
To me, it's just an implementation detail. My preferred implementation is to just call it again instead of having the code replicated. Anyway, this path can happen at most once for any QPDF object, and it's a very rare case.